src/neo/SmartContract/ApplicationEngine.Runtime.cs

```diff
@@ -187,6 +188,39 @@ protected internal void SendNotification(UInt160 hash, string eventName, Array s
             Notify?.Invoke(this, notification);
             notifications ??= new List<NotifyEventArgs>();
             notifications.Add(notification);
+
+            // React to transfer event
+
+            if (eventName == "Transfer")
+            {
+                if (state.Count < 3) throw new ArgumentException("Wrong Transfer definition");
+
+                var to = state[1].GetSpan();
+                if (to.Length != UInt160.Length) throw new ArgumentException("Wrong Transfer definition");
+
+                var contract = Snapshot.Contracts.TryGet(new UInt160(to));
+                if (contract == null) return;
+                if (!contract.Payable) throw new InvalidOperationException("Not payable contract");
+
+                // Validate
+
+                var method = contract.Manifest.Abi.GetMethod("_onPaymentReceived");
```

Comment:
May happen re-entrancy attack?
```
 Nep5A.Transfer -> `_onPaymentReceived` -> NepA.Transfer .......?  
```

Then the user's asset can be transfered

Reply:
Of course, the users need to be aware and use invocation counter